lucet-runtime/lucet-runtime-internals/src/future.rs

```diff
@@ -49,65 +54,67 @@ impl Vmctx {
-            State::Running { async_context } => {
-                if !async_context {
-                    panic!(TerminationDetails::BlockOnNeedsAsync)
-                }
+        // Wrap the future, so that we don't have to worry about sending back the return value
+        let ret = UnsafeCell::new(MaybeUninit::uninit());
+        let ret_ptr = ret.get();
+
+        let mut f = async move {
+            let ret = MaybeUninit::new(f.await);
+            // SAFETY: we are the only possible writer to ret_ptr, and the future must be polled to completion before this function returns
+            unsafe {
+                std::ptr::write(ret_ptr, ret);
             }
-            _ => unreachable!("Access to vmctx implies instance is Running"),
-        }
-        // Wrap the Output of `f` as a boxed ResumeVal. Then, box the entire
-        // async computation.
-        let f = Box::pin(async move { ResumeVal(Box::new(f.await)) });
-        // Change the lifetime of the async computation from `'a` to `'static.
-        // We need to lie about this lifetime so that `YieldedFuture` may impl
-        // `Any` and be passed through the yield. `Instance::run_async`
-        // rehydrates this lifetime to be at most as long as the Vmctx's `'a`.
-        // This is safe because the stack frame that `'a` is tied to gets
-        // frozen in place as part of `self.yield_val_expecting_val`.
-        let f = unsafe {
-            std::mem::transmute::<BoxFuture<'a, ResumeVal>, BoxFuture<'static, ResumeVal>>(f)
+            ResumeAsync
         };
-        // Wrap the computation in `YieldedFuture` so that
-        // `Instance::run_async` can catch and run it.  We will get the
-        // `ResumeVal` we applied to `f` above.
-        self.yield_impl::<YieldedFuture, ResumeVal>(YieldedFuture(f), false, false);
-        let ResumeVal(v) = self.take_resumed_val();
-        // We may now downcast and unbox the returned Box<dyn Any> into an `R`
-        // again.
-        *v.downcast().expect("run_async broke invariant")
-    }
-}
 
-/// A simple future that yields once. We use this to yield when a runtime bound is reached.
-///
-/// Inspired by Tokio's `yield_now()`.
-struct YieldNow {
-    yielded: bool,
-}
+        // We pin the future to the stack (a requirement for being able to poll the future).
+        // By pinning to the stack instead of using `Box::pin`, we avoid heap allocations for immediately-ready futures.
+        //
+        // SAFETY: We must uphold the invariants of Pin, namely that future does not move until it is dropped.
+        // By overriding the variable named `f`, it is impossible to access f again, except through the pinned reference.
+        let mut f = unsafe { Pin::new_unchecked(&mut f) };
 
-impl YieldNow {
-    fn new() -> Self {
-        Self { yielded: false }
-    }
-}
+        // Poll the future using the std::task::Context, or die if we aren't async
+        match &mut unsafe { self.instance_mut() }.state {
```

Comment:
Note: this is probably unsound, and needs to be a refcell - re-entrant calls to block_on would cause undefined behavior (two mutable references to state)